File: protocol-generator/dart/test/scope_algebra_test.dart
import 'package:entity_core_protocol/entity_core_peer.dart';
import 'package:test/test.dart';

/// The §5 scope algebra at 0.8.2.24/.25 — the two rulings that changed what an
/// existing matcher must answer, plus the call-site enumeration RULE F asks for.
///
/// UNIT-level on purpose. §5.5a attenuation and the id-scope/path-scope split are
/// decided inside pure functions whose inputs a wire drive cannot reach directly:
/// `arc-probe` can show a grant being honoured or refused, it cannot show WHICH
/// matcher decided. The pinned 778-check set has no vector on either.
void main() {
  const local = 'z6MkLocalPeerIdThatIsNotRealButIsLongEnoughToPassBase58Check';

  Scope sc(List<String> incl, [List<String> excl = const []]) => Scope(incl, excl);

  /// A grant whose four dimensions move independently.
  GrantRec grant({
    Scope? handlers,
    Scope? resources,
    Scope? operations,
    Scope? peers,
  }) =>
      GrantRec(handlers ?? sc(['*']), resources ?? sc(['*']),
          operations ?? sc(['*']), peers);

  group('the sentinel is scoped to path-scope (0.8.2.24 N2/N3)', () {
    // §5.4 at 0.8.2.24: "a capability carrying an unmatchable PATH-SCOPE pattern is
    // INVALID ... It does NOT reach `operations` or `peers` [MUST]".
    //
    // The un-scoped form shipped at 0.8.2.21 transcribed §5.2's exclude loop before
    // that loop grew its type dispatch, so an id-scope pattern was run through the
    // §5.4 PATH transforms purely to classify it and the whole dimension was then
    // denied on a property unrelated to whether the exclude carves anything out.

    test('an operations exclude that path-canonicalizes to the sentinel does NOT deny',
        () {
      // `*/apply` is an ordinary namespaced operation name. Under the id-scope
      // grammar it is a LITERAL that matches nothing — a non-match, never a fault;
      // under the §5.4 path transforms it canonicalizes to the sentinel. Before the
      // fix this denied EVERY operation. Over-denial, invisible on a well-formed
      // grant.
      expect(canonicalize(local, '*/apply'), equals(neverMatch));
      expect(matchesScope(local, 'get', sc(['*'], ['*/apply']), ScopeKind.id),
          isTrue);
    });

    test('the same exclude on a PATH-scope dimension still denies (0.8.2.21 intact)',
        () {
      // The control that says the fix SCOPED the guard rather than deleting it. An
      // unmatchable exclude left to the matcher is fail-OPEN — it carves out nothing
      // and the grant is silently wider than its author wrote — so on a path-scope
      // dimension it must still deny.
      expect(
          matchesScope(local, 'system/tree', sc(['*'], ['*/apply']), ScopeKind.path),
          isFalse);
    });

    test('a peers exclude naming a relative-traversal form does not deny', () {
      // The second id-scope dimension, driven independently: a claim about
      // `operations` alone would not say the guard moved off `peers` too.
      expect(canonicalize(local, '../elsewhere'), equals(neverMatch));
      expect(
          matchesScope(local, local, sc([local], ['../elsewhere']), ScopeKind.id),
          isTrue);
    });

    test('an ordinary MATCHABLE exclude still excludes on both scope kinds', () {
      // The FIXTURE control. Accept cases alone cannot tell a working matcher from
      // one that allows everything; deny cases alone cannot tell it from one that
      // denies everything.
      expect(matchesScope(local, 'get', sc(['*'], ['get']), ScopeKind.id), isFalse);
      expect(
          matchesScope(
              local, 'system/tree', sc(['*'], ['system/tree']), ScopeKind.path),
          isFalse);
    });
  });

  group('scope_subset is typed by scope kind (F50, ruled 0.8.2.16)', () {
    // §3.6's grammar binds the SCOPE TYPE, not one function: "An implementation on
    // the canonicalizing reading is non-conformant and MUST adopt the literal
    // matcher." F40 typed `matchesScope`; its §5.5a sibling was left on the path
    // matcher for all four dimensions.
    //
    // Driven through `grantSubset`, the public call site — a test of the private
    // helper would prove it CAN dispatch and say nothing about whether the four
    // dimensions NAME their kind.

    test('an operations include the ID matcher covers is a subset of a bare star',
        () {
      // The formalization's witness. Under the path matcher the parent canonicalizes
      // to `/{parent}/*` and the child to the sentinel, so the pair was refused.
      // FAIL-CLOSED, which is exactly why no hand-tried example found it: nothing is
      // over-granted, legitimate delegation is refused.
      expect(
          grantSubset(local, local, local, grant(operations: sc(['*/apply'])),
              grant(operations: sc(['*']))),
          isTrue);
    });

    test('the second witness: an absolute-looking operation name under a star', () {
      expect(
          grantSubset(local, local, local, grant(operations: sc(['/tree/get'])),
              grant(operations: sc(['*']))),
          isTrue);
    });

    test('the SAME two patterns on RESOURCES are still path-matched (refused)', () {
      // THE DIFFERENTIAL that attributes the two rows above to the scope TYPE rather
      // than to `grantSubset` having been loosened.
      expect(
          grantSubset(local, local, local, grant(resources: sc(['*/apply'])),
              grant(resources: sc(['*']))),
          isFalse);
      expect(
          grantSubset(local, local, local, grant(resources: sc(['/tree/get'])),
              grant(resources: sc(['*']))),
          isFalse);
    });

    test('an operations include no parent include covers is still refused', () {
      // The accept rows cannot distinguish "the id matcher was used" from "the
      // operations dimension stopped being checked".
      expect(
          grantSubset(local, local, local, grant(operations: sc(['put'])),
              grant(operations: sc(['get']))),
          isFalse);
    });

    test('a parent operations EXCLUDE not inherited by the child is refused', () {
      // The exclude arm of the subset test on the id dimension, independently: the
      // include arm alone would not say the exclude arm is typed.
      expect(
          grantSubset(local, local, local, grant(operations: sc(['*'])),
              grant(operations: sc(['*'], ['delete']))),
          isFalse);
      expect(
          grantSubset(local, local, local, grant(operations: sc(['*'], ['delete'])),
              grant(operations: sc(['*'], ['delete']))),
          isTrue);
    });
  });

  group('the sentinel guard is reached from every match decision (RULE F, K-6)', () {
    // 0.8.2.22: "a sentinel arm is a control-flow obligation, not a line ... the
    // guard MUST sit on every path that reaches the decision it protects." The `lean`
    // bypass was a GUARDED WRAPPER beside an UNGUARDED running matcher, with
    // attenuation calling the raw one.
    //
    // ALREADY SATISFIED BY CONSTRUCTION HERE, and these cases are the measurement
    // that says so rather than a claim. `matchesPattern` is the single definition and
    // the sentinel test is its FIRST statement, so the bypass shape cannot occur: the
    // six call sites in lib/ (its own recursion, `_covered`, `_coveredFrame`,
    // `effectiveTargets`'s caller-exclude arm, and `_scopeSubset`'s two path arms)
    // all reach the same guarded body.
    //
    // WHICH OF THESE THE GUARD IS LOAD-BEARING FOR WAS MEASURED, NOT ASSUMED.
    // Deleting the guard and re-running shows the pattern-operand half is unreachable
    // as a discriminator here: `matchesPattern(x, neverMatch)` can only be true when
    // `x == neverMatch`, because the sentinel is neither a bare star, nor a `/*/`
    // form, nor a `/*` suffix, so every other arm falls through to the literal
    // compare and answers false on its own. A case that passes with the guard deleted
    // is not a control for the guard — it is a control for the fall-through.

    test('refuses the sentinel as the PATH operand', () {
      // The arm that makes the guard load-bearing: a bare star returns true for
      // anything, so safety must not rest on a value merely LOOKING unmatchable.
      expect(matchesPattern(neverMatch, '*'), isFalse);
    });

    test('refuses the sentinel against ITSELF', () {
      // The case a guard written as `path != pattern` would pass, and the only input
      // where the PATTERN-operand half decides anything.
      expect(matchesPattern(neverMatch, neverMatch), isFalse);
    });

    test('the guard is reached through ATTENUATION, not only through matchesScope',
        () {
      // The `lean` defect in this peer's own shape. The parent include is `/*` and
      // NOT a bare `*` deliberately: a bare star canonicalizes to `/{parent}/*`,
      // which refuses a sentinel child by prefix alone and would make this case pass
      // with the guard deleted. `/*` is already absolute, survives canonicalization
      // unchanged, and its `/*` suffix arm covers any path beginning with a slash —
      // including the sentinel. Only the guard refuses it.
      expect(
          grantSubset(local, local, local, grant(resources: sc(['../escape'])),
              grant(resources: sc(['/*']))),
          isFalse);
    });
  });
}

File: protocol-generator/php/tests/ScopeAlgebraTest.php
<?php

declare(strict_types=1);

namespace EntityCore\Tests;

use EntityCore\Capability;
use EntityCore\Ecf;
use EntityCore\EcfMap;
use EntityCore\Entity;
use EntityCore\Verdict;
use PHPUnit\Framework\TestCase;

/**
 * The 0.8.2.20 -> 0.8.2.25 scope-algebra units: §3.3's effective-target projection,
 * §6.3's handler-level path check, §5.2's PATH-SCOPED sentinel guard, and §5.5a's
 * scope-kind-typed subset check.
 *
 * These are the pure predicates. The §3.3 LADDER (which arm answers which code) and the
 * operation-before-resource ordering are driven over a real socket in
 * {@see TreeLadderTest}, because both are properties of the handler's control flow rather
 * than of any one predicate.
 */
final class ScopeAlgebraTest extends TestCase
{
    private const LOCAL = '2KHoAk7A5JmhygZJAdBua8iRD1CnBoJRfUBHgZeXNRTeFg';
    private const PATTERN = '/2KHoAk7A5JmhygZJAdBua8iRD1CnBoJRfUBHgZeXNRTeFg/system/tree';
    private const COVERED = '/2KHoAk7A5JmhygZJAdBua8iRD1CnBoJRfUBHgZeXNRTeFg/system/type/qA';
    private const OUTSIDE = '/2KHoAk7A5JmhygZJAdBua8iRD1CnBoJRfUBHgZeXNRTeFg/secrets/qB';

    /**
     * A `resource` map: `{targets: [...]}` plus `{exclude: [...]}` when non-empty.
     *
     * @param list<string> $targets
     * @param list<string> $excl
     */
    private static function resource(array $targets, array $excl = []): EcfMap
    {
        return $excl === []
            ? Ecf::map('targets', $targets)
            : Ecf::map('targets', $targets, 'exclude', $excl);
    }

    private static function execWith(string $operation, ?EcfMap $resource): Entity
    {
        $fields = ['operation', $operation, 'uri', self::PATTERN];
        if ($resource !== null) {
            $fields[] = 'resource';
            $fields[] = $resource;
        }
        return Entity::make('system/protocol/execute', Ecf::map(...$fields));
    }

    /**
     * A scope map. `include`/`exclude` are BOTH emitted so an empty include is
     * distinguishable from an absent one — §5.2 makes an empty include a legal grant
     * shape that denies everything, and a fixture that dropped the key would test the
     * wrong thing.
     *
     * @param list<string> $incl
     * @param list<string> $excl
     */
    private static function scope(array $incl, array $excl = []): EcfMap
    {
        return Ecf::map('include', $incl, 'exclude', $excl);
    }

    /**
     * @param array{0:list<string>,1:list<string>} $handlers
     * @param array{0:list<string>,1:list<string>} $operations
     * @param array{0:list<string>,1:list<string>} $resources
     */
    private static function grant(array $handlers, array $operations, array $resources): EcfMap
    {
        return Ecf::map(
            'handlers', self::scope($handlers[0], $handlers[1]),
            'operations', self::scope($operations[0], $operations[1]),
            'resources', self::scope($resources[0], $resources[1]),
        );
    }

    private static function token(EcfMap ...$grants): Entity
    {
        return Entity::make('system/capability/token', Ecf::map('grants', \array_values($grants)));
    }

    // ── §3.3 effective targets (0.8.2.20/.21, N11) ────────────────────────────────

    public function testEffectiveTargetsIsANonLossyProjection(): void
    {
        // ABSENT resource — the two empties must be TELLABLE APART (N11). A projection
        // that answered `[]` here would delete the discriminator before any handler could
        // read it, and `get`'s absent arm (a root listing) and its self-excluded arm
        // (400 path_required) would collapse into one.
        self::assertNull(
            Capability::effectiveTargets(self::LOCAL, self::execWith('get', null)),
            'an absent resource must be reported as absent, not as an empty list',
        );

        // PRESENT, every target carved out by the caller's own exclude: NOT null, and an
        // empty survivor list. This is the other half of the discriminator.
        self::assertSame(
            [],
            Capability::effectiveTargets(self::LOCAL, self::execWith('get', self::resource(['a/b'], ['a/*']))),
            'a present resource whose every target is excluded is [] and not null',
        );

        // SURVIVORS COME BACK IN THE CALLER'S OWN SPELLING, not canonicalized (0.8.2.21).
        // The value flows on to the store lookup, which canonicalizes for itself; handing
        // back a canonical form here would double-canonicalize a relative target.
        self::assertSame(
            ['x/one'],
            Capability::effectiveTargets(self::LOCAL, self::execWith('get', self::resource(['x/one', 'x/two'], ['x/two']))),
            'the survivor keeps the caller own spelling',
        );

        // THE SELECTION MUST (F84): the survivor is the one the exclude LEFT, never
        // targets[0]. That is the whole point — a handler that counts the effective list
        // and then indexes the raw targets has the arithmetic right and reads a path no
        // authorization covered.
        self::assertSame(
            ['x/two'],
            Capability::effectiveTargets(self::LOCAL, self::execWith('get', self::resource(['x/one', 'x/two'], ['x/one']))),
            'the survivor is x/two, not targets[0]',
        );

        // THE CALLER-EXCLUDE ARM IS FAIL-OPEN ON AN UNMATCHABLE PATTERN. §5.4 rules it
        // separately from the grant arm: canonicalize answers the sentinel,
        // matchesPattern then answers false, and the target simply SURVIVES. Inherited,
        // not restated — this assertion exists so the inheritance is measured rather than
        // assumed.
        self::assertSame(
            ['x/one'],
            Capability::effectiveTargets(self::LOCAL, self::execWith('get', self::resource(['x/one'], ['../nope']))),
            'an unmatchable CALLER exclude carves out nothing',
        );

        // A PRESENT-BUT-ILL-TYPED `targets` IS **PRESENT**, with an empty survivor list.
        // Reporting it absent would serve the WIDER absent-case answer to a request that
        // named a resource — N11's own defect one field over.
        self::assertSame(
            [],
            Capability::effectiveTargets(self::LOCAL, self::execWith('get', Ecf::map('targets', 7))),
            'an ill-typed targets is still a PRESENT resource',
        );
    }

    // ── §6.3 check_path_permission (0.8.2.20/.22/.23) ─────────────────────────────

    public function testCheckPathPermissionConsultsThreeDimensions(): void
    {
        // The ACCEPT case first, and it is the one that validates the FIXTURE: a suite
        // built only from deny cases is indistinguishable from one asserting
        // False === False, which a mis-built grant fixture guarantees for free.
        $tok = self::token(self::grant([['system/tree'], []], [['get'], []], [['system/type/*'], []]));
        self::assertTrue(
            Capability::checkPathPermission(self::LOCAL, 'get', self::COVERED, $tok, self::PATTERN),
            'a grant covering handler+operation+resource must ALLOW',
        );
        // One deny per DIMENSION — a single deny cannot distinguish "the predicate checks
        // the dimension I care about" from "the predicate denies".
        self::assertFalse(
            Capability::checkPathPermission(self::LOCAL, 'get', self::OUTSIDE, $tok, self::PATTERN),
            'RESOURCES: a path outside resources.include must DENY',
        );
        self::assertFalse(
            Capability::checkPathPermission(self::LOCAL, 'put', self::COVERED, $tok, self::PATTERN),
            'OPERATIONS: an operation outside operations.include must DENY',
        );
        self::assertFalse(
            Capability::checkPathPermission(self::LOCAL, 'get', self::COVERED, $tok, '/x/system/other'),
            'HANDLERS: a handler pattern outside handlers.include must DENY',
        );

        // AN EMPTY `resources.include` IS A LEGAL GRANT SHAPE and denies EVERY path
        // (§5.2: handlers that touch no tree paths). `covered` over an empty include list
        // is false, which is what that note says it should be.
        $empty = self::token(self::grant([['system/tree'], []], [['get'], []], [[], []]));
        self::assertFalse(
            Capability::checkPathPermission(self::LOCAL, 'get', self::COVERED, $empty, self::PATTERN),
            'an empty resources.include denies every path',
        );

        // A GRANT EXCLUDE COVERING THE SUBJECT DENIES, even though the include covers it.
        $excl = self::token(self::grant([['system/tree'], []], [['get'], []], [['system/type/*'], ['system/type/qA']]));
        self::assertFalse(
            Capability::checkPathPermission(self::LOCAL, 'get', self::COVERED, $excl, self::PATTERN),
            'a grant exclude covering the subject denies',
        );

        // A MALFORMED PATH canonicalizes to the §5.4 sentinel, which matches no grant, so
        // it falls through to DENY rather than being matched against anything — including
        // against a grant whose include is a bare star.
        $star = self::token(self::grant([['system/tree'], []], [['get'], []], [['*'], []]));
        self::assertTrue(
            Capability::checkPathPermission(self::LOCAL, 'get', self::COVERED, $star, self::PATTERN),
            'control: the bare-star grant does allow an ordinary path',
        );
        self::assertFalse(
            Capability::checkPathPermission(self::LOCAL, 'get', '../nope', $star, self::PATTERN),
            'a path that canonicalizes to NEVER_MATCH denies',
        );

        // THE OPERATIONS DIMENSION IS ID-SCOPE, NOT PATH-SCOPE (F40). A path-form pattern
        // in `operations` is matched as a LITERAL string: a non-match, never a fault, and
        // never a canonicalizing match against an unrelated operation name.
        $idScoped = self::token(self::grant([['system/tree'], []], [['/*/get'], []], [['*'], []]));
        self::assertFalse(
            Capability::checkPathPermission(self::LOCAL, 'get', self::COVERED, $idScoped, self::PATTERN),
            'operations is id-scope: a path-form pattern is a literal and does not match get',
        );
    }

    // ── §5.2 the sentinel is scoped to PATH-SCOPE (0.8.2.24, N2/N3) ───────────────

    public function testSentinelGuardIsPathScopeOnly(): void
    {
        // AN ID-SCOPE EXCLUDE THAT PATH-CANONICALIZES TO THE SENTINEL MUST NOT DENY THE
        // WHOLE DIMENSION. `*/apply` is an ordinary namespaced operation name and a
        // literal under the id-scope grammar; putting it through the §5.4 transforms
        // purely to classify it produced the sentinel and denied EVERY operation.
        // Over-denial, and invisible on any well-formed grant — which is why this needs a
        // test rather than a reading.
        $tok = self::token(self::grant([['*'], []], [['*'], ['*/apply']], [['*'], []]));
        self::assertSame(
            Verdict::Allow,
            Capability::checkPermission(self::LOCAL, self::LOCAL, self::execWith('get', null), $tok, self::PATTERN),
            'an id-scope exclude that path-canonicalizes to the sentinel must not deny the dimension',
        );

        // THE PATH-SCOPE ARM STILL DENIES — the control that says the guard was SCOPED
        // rather than DELETED. An unmatchable exclude on `resources` excludes everything
        // (0.8.2.21), because there the sentinel means the author wrote a carve-out that
        // carves nothing and the grant is silently wider than written.
        $res = self::token(self::grant([['*'], []], [['*'], []], [['*'], ['../nope']]));
        self::assertSame(
            Verdict::Deny,
            Capability::checkPermission(
                self::LOCAL,
                self::LOCAL,
                self::execWith('get', self::resource(['system/type/qA'])),
                $res,
                self::PATTERN,
            ),
            'an unmatchable PATH-SCOPE exclude still denies (0.8.2.21)',
        );

        // And the same guard on the HANDLERS dimension, the other path-scope one.
        $hand = self::token(self::grant([['*'], ['../nope']], [['*'], []], [['*'], []]));
        self::assertSame(
            Verdict::Deny,
            Capability::checkPermission(self::LOCAL, self::LOCAL, self::execWith('get', null), $hand, self::PATTERN),
            'an unmatchable handlers exclude still denies (handlers is path-scope)',
        );
    }

    // ── §5.5a scope_subset is typed by scope kind (F50 / 0.8.2.16, K-7) ───────────

    /**
     * @param array{0:list<string>,1:list<string>} $childOps
     * @param array{0:list<string>,1:list<string>} $parentOps
     * @param array{0:list<string>,1:list<string>} $childRes
     * @param array{0:list<string>,1:list<string>} $parentRes
     */
    private static function subset(array $childOps, array $parentOps, array $childRes = [['*'], []], array $parentRes = [['*'], []]): bool
    {
        $parse = static function (EcfMap $g): array {
            $tok = self::token($g);
            return Capability::grantsOfToken($tok)[0];
        };
        return Capability::grantSubset(
            self::LOCAL,
            self::LOCAL,
            self::LOCAL,
            $parse(self::grant([['*'], []], $childOps, $childRes)),
            $parse(self::grant([['*'], []], $parentOps, $parentRes)),
        );
    }

    public function testScopeSubsetIsTypedByScopeKind(): void
    {
        // THE DIFFERENTIAL. `entity-core-formalization` measured 2 of 64 include pairs
        // disagreeing between the two readings, FAIL-CLOSED, with a 16-pair control
        // alphabet reporting 0 — which is why every hand-tried example missed it. Both
        // witnesses (`*/apply`, `/tree/get`) are OPERATIONS patterns, which §3.6 types as
        // id-scope: under the literal matcher a bare-star parent covers any child
        // pattern, and under the canonicalizing matcher the child canonicalizes to the
        // sentinel (or to an absolute path the parent's peer-framed star cannot cover)
        // and the subset is wrongly REFUSED.
        self::assertTrue(
            self::subset([['*/apply'], []], [['*'], []]),
            'id-scope: a namespaced operation child is covered by a bare-star parent',
        );
        self::assertTrue(
            self::subset([['/tree/get'], []], [['*'], []]),
            'id-scope: a path-form operation child is covered by a bare-star parent',
        );

        // THE CONTROL ALPHABET — the pairs that agree under both readings. Without these
        // the two assertions above are equally explained by a subset check that has
        // stopped checking anything.
        self::assertTrue(self::subset([['get'], []], [['*'], []]), 'control: get is covered by a bare-star parent');
        self::assertFalse(self::subset([['put'], []], [['get'], []]), 'control: put is NOT covered by a get-only parent');

        // A WIDENING ON THE PATH-SCOPE DIMENSION IS STILL REFUSED — the kind parameter
        // selected the matcher, it did not disable the check.
        self::assertFalse(
            self::subset([['get'], []], [['get'], []], [['*'], []], [['system/type/*'], []]),
            'path-scope: a bare-star child is NOT covered by a narrower parent',
        );

        // A PARENT EXCLUDE MUST BE INHERITED BY THE CHILD, on the id arm too.
        self::assertFalse(
            self::subset([['get'], []], [['*'], ['put']]),
            'id-scope: a child that does not inherit the parent exclude is refused',
        );
    }
}
